liquidctl/driver/coolit.py

```diff
@@ -0,0 +1,415 @@
+
+Supported devices
+-----------------
+
+ - Corsair H110i GT
+
+Supported features
+------------------
+
+ - general monitoring
+ - pump speed control
+ - fan speed control
+
+Copyright Roberto Marques, Serphentas, and other contributors
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+import logging
+import re
+
+from enum import Enum, unique
+
+from liquidctl.driver.usb import UsbHidDriver
+from liquidctl.keyval import RuntimeStorage
+from liquidctl.util import clamp, fraction_of_byte, u16le_from, normalize_profile
+
+
+LOGGER = logging.getLogger(__name__)
+
+_REPORT_LENGTH = 64
+_PUMP_INDEX = 0x02
+
+_COMMAND_FIRMWARE_ID = 0x01
+_COMMAND_TEMP_READ = 0x0E
+_COMMAND_FAN_SELECT = 0x10
+_COMMAND_FAN_MODE = 0x12
+_COMMAND_FAN_FIXED_PWM = 0x13
+_COMMAND_FAN_FIXED_RPM = 0x14
+_COMMAND_FAN_READ_RPM = 0x16
+_COMMAND_FAN_MAX_RPM = 0x17
+_COMMAND_FAN_RPM_TABLE = 0x19
+_COMMAND_FAN_TEMP_TABLE = 0x1A
+
+_OP_CODE_WRITE_ONE_BYTE = 0x06
+_OP_CODE_READ_ONE_BYTE = 0x07
+_OP_CODE_WRITE_TWO_BYTES = 0x08
+_OP_CODE_READ_TWO_BYTES = 0x09
+_OP_CODE_WRITE_THREE_BYTES = 0x0A
+_OP_CODE_READ_THREE_BYTES = 0x0B
+
+_PROFILE_LENGTH = 5
+_CRITICAL_TEMPERATURE = 60
+
+_PUMP_DEFAULT_QUIET = [0x2E, 0x09]
+_PUMP_DEFAULT_EXTREME = [0x86, 0x0B]
+
+_SEQUENCE_MIN = 1
+_SEQUENCE_MAX = 255
+
+
+@unique
+class _FanMode(Enum):
+    FIXED_DUTY = 0x02
+    FIXED_RPM = 0x04
+    CUSTOM_PROFILE = 0x0E
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to FIXED_DUTY for _FanMode(%s)", value)
+        return _FanMode.FIXED_DUTY
+
+
+@unique
+class _PumpMode(Enum):
+    QUIET = 0x08
+    EXTREME = 0x0C
+
+    @classmethod
+    def _missing_(cls, value):
+        LOGGER.debug("falling back to QUIET for _PumpMode(%s)", value)
+        return _PumpMode.QUIET
+
+
+def _sequence():
+    """Return a generator that produces valid protocol sequence numbers.
+
+    Sequence numbers are random.
+    """
+    num = 1
+    while True:
+        yield (num % 31) + 1
+        num += 1
+
+
+def _prepare_profile(original):
+    clamped = ((temp, clamp(duty, 0, 100)) for temp, duty in original)
+    normal = normalize_profile(clamped, _CRITICAL_TEMPERATURE)
+    missing = _PROFILE_LENGTH - len(normal)
+    if missing < 0:
+        raise ValueError(f"Too many points in profile (remove {-missing})")
+    if missing > 0:
+        normal += missing * [(_CRITICAL_TEMPERATURE, 100)]
+    return normal
+
+
+def _quoted(*names):
+    return ", ".join(map(repr, names))
+
+
+class CoolitDriver(UsbHidDriver):
+    """liquidctl driver for Corsair H110i GT cooler"""
+
+    _MATCHES = [
+        (
+            0x1B1C,
+            0x0C04,
+            "Corsair H110i GT (experimental)",
+            {"has_pump": True, "fan_count": 2, "rgb_fans": False},
+        ),
+    ]
+
+    def __init__(self, device, description, fan_count, rgb_fans, **kwargs):
+        super().__init__(device, description, **kwargs)
+        self._component_count = 1 + fan_count * rgb_fans
```

Comment:
This field and the `rgb_fans` argument are currently unused. How do you plan to use them in the future? And are there really any devices from this generation with RGB fans?

Reply:
Per [product page](https://www.corsair.com/us/en/p/cpu-coolers/cw-9060026-ww/hydro-series-h110i-280mm-extreme-performance-liquid-cpu-cooler-cw-9060026-ww#tab-overview), there is at least one RGB channel that can be controlled. This feature is not currently supported, so I think it can be left for future implementation.

Reply:
Ack.